### ISSUE TITLE
The workflow you've loaded may have connection/linking data that could be fixed. Open fixer  |  Fix in place

### ISSUE THREAD

**Issue by DarkAlchy** [chrisgoringe/cg-use-everywhere#51]
![image](https://github.com/chrisgoringe/cg-use-everywhere/assets/10561690/ac6b916e-4866-47fb-86e0-c63d06f46dd2)

I get that every single time I load a workflow with this in it.  Any ideas?

**Comment by chrisgoringe** [comment]
This is another custom node plugin (pythonssss? Maybe?) - you can ignore it. I'll see if I can work it out with the other developer!

**Comment by DarkAlchy** [comment]
> This is another custom node plugin (pythonssss? Maybe?) - you can ignore it. I'll see if I can work it out with the other developer!

Would be great and I love this as I have been shouting for a long time now that we needed nodes like this to get rid of the noodles only this one went one better than I was talking about.   Fantastic work with this one crazy other plugin being triggered.  pythonssss is too valuable to remove too.

**Comment by chrisgoringe** [comment]
Reopening this to prompt me to work out out with pythonsssss

**Comment by DarkAlchy** [comment]
My fingers are crossed.

**Comment by aegis72** [comment]
Same here, but the nodes are amazing, especially as I look to increasingly make my workflow into modular templates! Too cool for words. Also, I can confirm it works with Efficiency Nodes: 
![image](https://github.com/chrisgoringe/cg-use-everywhere/assets/118572301/819d97ca-a33d-4af3-85aa-2403c8d3ea9b)

**Comment by DarkAlchy** [comment]
Yep, all my workflows has this now, but it is so annoying that the popup never goes away on its own, I have to click it.  I hope they work it out.

**Comment by chrisgoringe** [comment]
thanks, rg. It would be a big help if anyone who has a workflow where this happens could post it here, so I can investigate whether there is a UE bug or a situation that could be handled better.

**Comment by chrisgoringe** [comment]
@rgthree - can you give me the tl;dr version of what you're checking for? UE nodes don't modify the underlying graph, but hijack the graph_to_prompt method to modify what gets sent in the prompt.

**Comment by rgthree** [comment]
Totally. Also, note, most ppl have a ton of nodes. This may not actually be a UE issue, especially if it's not changing the graph data in the UI.

Okay, so when a workflow is loaded (manually, or dragging a JSON or Image into the window) it checks for the _live_ `app.graph` for [bad links](https://github.com/rgthree/rgthree-comfy/blob/65832bcc83880e8ad5c57cef04b4754673dba971/ts/link_fixer.ts#L38). (If there was an error loading the workflow, then it will check the SerializedGraph directly from the workflow instead of the live ComfyUI graph. The code is messy to accommodate/normalize both graph data types.)

Anyway, the routine goes over the `graph.links` list, and verifies that the data in every defined `link` can be corroborated with the nodes it says it's linking. That is:

1. that the origin and target nodes for the list even exist in the graph nodes list
2. that the origin node's `outputs[link.origin_slot].links` has that link id in its list
3. that the target node's `inputs[link.target_slot].link` has that link id defined there, etc.

And than take a action from several possibilities depending on what's identified above.

The source is at [rgthree-comfy/blob/main/ts/link_fixer.ts](https://github.com/rgthree/rgthree-comfy/blob/65832bcc83880e8ad5c57cef04b4754673dba971/ts/link_fixer.ts#L38, though there is a lot happening in there, the start of the iteration would be [further down](https://github.com/rgthree/rgthree-comfy/blob/65832bcc83880e8ad5c57cef04b4754673dba971/ts/link_fixer.ts#L220)

**Comment by DarkAlchy** [comment]
@rgthree With those revelations what is the everywhere node not doing that is failing those checks to make the pop up happen?

**Comment by chrisgoringe** [comment]
@rgthree - in link_fixer.js line 240 I think the log message is incorrect? `origin_node` and `target_node` used inconsistently?:

```
                logger.log(`${link.id} is def invalid; BOTH origin node ${link.origin_id} ${originNode ? "is removed" : `doesn\'t have ${link.id}`} and ${link.origin_id} target node ${link.target_id ? "is removed" : `doesn\'t have ${link.id}`}.`);
```

maybe should read

```
                logger.log(`${link.id} is def invalid; BOTH origin node ${link.origin_id} ${!originNode ? "is removed" : `doesn\'t have ${link.id}`} and ${link.target_id} target node ${!target_node ? "is removed" : `doesn\'t have ${link.id}`}.`);
```

**Comment by chrisgoringe** [comment]
AHA!

UE doesn't modify the graph, but what gets serialized when you save a workflow...? Looks like the workflow is being saved with UE pseudo-links included, instead of the graph being saved.

I've created a new issue #65 to track this, but I think it'll fix the problems here.

**Comment by chrisgoringe** [comment]
OK - this is fixed in 4.3.8. 

*new* saved workflows will no longer have the UE links included, and should not trigger the warning.
*existing* workflows (saved with UE before 4.3.8) will still have them; you can use rgthree's auto cleanup to remove them, and then resave, and all should be well.

If you find this warning recurs after that, please reopen this issue (or open a new one).

**Comment by rgthree** [comment]
Awesome! (And, @chrisgoringe , I think you're right about the message in link_fixer).
